crates/wasmparser/src/binary_reader.rs

```diff
@@ -242,78 +300,458 @@ impl<'a> BinaryReader<'a> {
+                        x
+                    ),
+                    self.original_position() - 1,
+                ))
+            }
         })
     }
 
-    fn read_export_types(&mut self) -> Result<Box<[ExportType<'a>]>> {
-        let pos = self.original_position();
-        let exports_len = self.read_var_u32()? as usize;
-        if exports_len > MAX_WASM_EXPORTS {
-            return Err(BinaryReaderError::new("exports size is out of bound", pos));
-        }
-        (0..exports_len).map(|_| self.read_export_type()).collect()
+    pub(crate) fn read_interface_type(&mut self) -> Result<InterfaceType> {
+        let position = self.position;
+        Ok(match self.read_u8()? {
+            0x7f => InterfaceType::Unit,
+            0x7e => InterfaceType::Bool,
+            0x7d => InterfaceType::S8,
+            0x7c => InterfaceType::U8,
+            0x7b => InterfaceType::S16,
+            0x7a => InterfaceType::U16,
+            0x79 => InterfaceType::S32,
+            0x78 => InterfaceType::U32,
+            0x77 => InterfaceType::S64,
+            0x76 => InterfaceType::U64,
+            0x75 => InterfaceType::F32,
+            0x74 => InterfaceType::F64,
+            0x73 => InterfaceType::Char,
+            0x72 => InterfaceType::String,
+            x if x >= 0x80 => {
+                self.position = position;
+                InterfaceType::Compound(self.read_var_u32()?)
+            }
+            x => {
+                return Err(BinaryReaderError::new(
+                    format!("invalid leading byte (0x{:x}) for interface type", x),
+                    self.original_position() - 1,
+                ))
+            }
+        })
+    }
+
+    pub(crate) fn read_compound_type(&mut self, leading: u32) -> Result<CompoundType<'a>> {
+        Ok(match leading {
+            0x71 => CompoundType::Record(
+                (0..self.read_var_u32()?)
+                    .map(|_| Ok((self.read_string()?, self.read_interface_type()?)))
+                    .collect::<Result<_>>()?,
+            ),
+            0x70 => {
+                let cases = (0..self.read_var_u32()?)
+                    .map(|_| Ok((self.read_string()?, self.read_interface_type()?)))
+                    .collect::<Result<_>>()?;
+
+                let default = match self.read_u8()? {
+                    0x00 => None,
+                    0x01 => Some(self.read_var_u32()?),
+                    x => {
+                        return Err(BinaryReaderError::new(
+                            format!("invalid leading byte (0x{:x}) for variant default", x),
+                            self.original_position() - 1,
+                        ))
+                    }
+                };
+
+                CompoundType::Variant { cases, default }
+            }
+            0x6f => CompoundType::List(self.read_interface_type()?),
+            0x6e => CompoundType::Tuple(
+                (0..self.read_var_u32()?)
```

Comment:
I'm onboard with changing these `Box<[T]>` fields to provide sub-readers specific to what's being read, although I agree that it's probably a pretty big thing to implement and perhaps best done with a follow-up PR.

In the meantime, perhaps places where we're doing these vectored reads could check limits, although we've, so far, checked limits in the validator and not in the parser.

Reply:
Yeah definitely don't implement a fancier sub-reader pattern here. Unfortunately though implementing a sub-reader pattern is also not great in `wasmparser` today. Ideally you want to entirely skip the bytes that your sub-reader is parsing, but the way `wasmparser` is structured we'd have to do a "structured skip" of all the components, and then afterwards create a reader around those bytes to hand-off. This is distinct from function bodies where they're prefixed with the size, in bytes, so the skip is easy (e.g. we can do an unstructured skip).

Really fixing this at the `wasmparser` level would involve not having methods like this but rather a stateful-form of decoder where you're required to consume the decoder before drop or something like that. That's an even thornier API-design question and definitely not worth getting into here.

Basically I'm just saying that the naive approach of what `BrTable` does today, for example, by it just skips over the input bytes to continue parsing, is probably not a great idea and one we don't want to bake deeply into `wasmparser`. I don't know what really to do differently, though. It's never really been a big enough problem to cause issues.

Finally though we can't forget about this. I would be worried about this pattern allocating `self.read_var_u32()?` items worth of space to get filled in by the iterator which is something we don't want as that's attacker-controlled-memory-allocation. I suspect that the pervasive collect-into-`Result` means that the pre-allocation isn't happening, but then that's also arguably worse because parsing is slower than it needs to be with reallocation.

Anyway, this is an unfortunate problem...

Reply:
I can, at least for now, impose limit checks like we do in `read_br_table` so that the parser isn't susceptible to excessively large allocations from parsing a component.